src/K.hs

```diff
@@ -73,85 +69,74 @@ makekSpec sources _ behaviour =
     else Bad (nowhere, "No storagelayout found")
 
 kCalldata :: Interface -> String
-kCalldata (Interface a b) =
+kCalldata (Interface a args) =
   "#abiCallData("
   <> show a <> ", "
-  <> (case b of
-       [] -> ".IntList"
-       args ->
-         intercalate ", " (fmap (\(Decl typ varname) -> "#" <> show typ <> "(" <> kVar varname <> ")") args))
+  <> if null args then ".IntList"
+     else intercalate ", " (fmap (\(Decl typ varname) -> "#" <> show typ <> "(" <> kVar varname <> ")") args)
   <> ")"
 
-kStorageName :: TStorageItem a -> String
-kStorageName (DirectInt _ name)    = kVar name
-kStorageName (DirectBool _ name)   = kVar name
-kStorageName (DirectBytes _ name)  = kVar name
-kStorageName (MappedInt _ name ixs) = kVar name <> "_" <> intercalate "_" (NonEmpty.toList $ fmap kExpr ixs)
-kStorageName (MappedBool _ name ixs) = kVar name <> "_" <> intercalate "_" (NonEmpty.toList $ fmap kExpr ixs)
-kStorageName (MappedBytes _ name ixs) = kVar name <> "_" <> intercalate "_" (NonEmpty.toList $ fmap kExpr ixs)
+kStorageName :: TStorageItem a -> When -> String
```

Comment:
Not as it stands, but I have been thinking the same thing and at one point toyed with such a solution.

The issue is that at the moment, the `When` is stored in the `TEntry`. I've essentially made the choice that `TStorageItem`s only know whether they are explicitly timed or not (because we need to propagate this information to the indices), but it is not until when we include a storage item in an expression that we decide whether we want to reference its prestate value or its poststate value (or neither, when that's possible).

But I believe it would be entirely possible to move the `When`/`Time t` into the storage item. I probably had a reason for abandoning that approach but I suspect that now that the rest of the code base has stabilized it would be relatively easy. Do we want this? Reasons for: name generation becomes safer and less awkward. Reasons against: a storage item doesn't only have one state, it has many states, so we shouldn't say that `pre(x)`, `post(x)` and `x` are different items in storage.

Reply:
I think I agree, exposing the Time on `TStorageItem` makes sense

Reply:
For reference, this was more difficult than we thought and we decided to just ignore it and merge the existing stuff for now.